dae/dae/dae.1
.TH DAE 1

.SH NAME
dae \-  
.SH SYNOPSIS
dae [<audiofile.wav>]

.SH OPTIONS
Dae currently takes no options.

.SH DESCRIPTION
Dae, THE DIGITAL AUDIO EDITOR

.SH INTRODUCTION

Dae was written to be a simple to use digital audio editor which work well
with speech under Linux.
Dae is specifically a console program and should NOT be run from a graphical
interface like Gnome terminal. To open a console under a graphical based 
Linux system, try Ctrl+Alt+F2 or any other function key (F1-F6) 
to login on a pure console.
It was designed to work like a conventional tape recorder with the added
features of removing and inserting bloks of data.
Data can also be inserted from an existing audio file.
If the record option is chosen when in the middle of a file, the recording
is made and afterwards, inserted into the file at the point where the record
option was pressed.

.SH STARTING THE PROGRAM
Dae is started from a console and takes one parameter, the existing audio
file that one wants to edit.
If dae is called without a file name, it creates an empty file called
unknown.wav.
dae myfile.wav
will start dae, load myfile.wav and start to play the file.
A backup of myfile.wav is made as myfile.wav~ before the file is loaded.
If myfile.wav does not exist, it is created as an empty file containing one
second silence.

The current version of dae works reliably on .wav files, so if you want to
edit a .mp3 or .ogg, first convert it into a .wav and use dae on the .wav.
The easiest way to do this, is to use sox with its mp3 and ogg support
package installed.
E.G. sox somefile.mp3 somefile.wav
To convert the edited .wav back to mp3 use lame.
To convert back to .ogg, use oggenc.

.SH NAVIGATING IN DAE
The file starts to play when it is loaded.
The following keys can be used to navigate back and forth while the file is
playing:
f1: stop, stops the playback of the file.
f2: play, starts playback at the current position.
f3: fast rewind, jumps back one minute in the audio.  It can be used either
while the file is playing or when playback is stopped.
f4: fast forward, jumps forward one minute in the audio.  It can also be
used while the file is playing or when it is stopped.
left-arrow: rewind, jumps back five seconds in the file and can be used like
fast forward and fast rewind.  Note that holding down the left-arrow is not
recommended as it does not really work well.
right-arrow: forward, toggles the speed of playback between the normal speed
and four times the speed of normal playback.  It can be used while the file
is playing or when the file is stopped.
l: Toggles lowering of playback speed to 70 percent of normal speed.
This is usefull when marking out small portions of audio.
home: go to start of the file.  If the file was playing when this key is
pressed, the file will start playing from the beginning.
f12: go to end of file, goes to the end of the audio and stops playback,
showing the position in seconds.  Fractions of seconds are also shown.

w: where am I, shows the current position in the file and can always be
used.  The position is shown in seconds with a fraction part.
j: jump to, prompts for a position in seconds and jumps to that position in
the file.

.SH INSERTING AUDIO FROM A FILE
Press f11 to stop playback and select a file to be inserted.
The pickafile file browser with is part of the kies software, is called up
if it is installed.  Use the up and down arrows to select the file you want
to insert and press enter to start inserting.
If pickafile is not available, you will be prompted to type the name of the
desired file.
Note: currently only .wav files can be reliably inserted.
The .wav file that you insert does not have to be in the same format, IE,
number of channels, sampeling rate etc.
It would be converted into the format of the file you are editing
automatically.
Note that the insertion takes quite long.

.SH RECORDING WITH DAE
Press f10 to start recording at the current position in the file.  You will
now see "recording".  You are now
recording from the default selected input audio device which could be a
microphone or line in.
While you are recording, you have the following options.
f1 to stop recording
f2 to resume recording and q to stop recording and take you back to your
original file.
After you have pressed q as described above, allow for some time for the
content to be inserted into the file.  It uses the same routine as for
inserting from a file for this task.

.SH WORKING WITH PARTS OF THE AUDIO FILE
The following tasks can be performed on a part of the audio file which you
are editing:
Play a part of the file, remove a part of the file and write a part of a
file to an external file.
Before any of these tasks can be done, the part of the file you want to work
with needs to be identified.
This is done by marking the positions of the start and end of the part you
want to work with.  Dae calles the part of the file thus marked, a block.
Press f5 to mark the start of the block and press f6 to mark the end of the
block.  You do not have to mark the start first, if it is conveniant you can
first mark the end and mark the start later.  You must have both a start and
end position marked before any task can be done which requires a block to
exist.
To play the marked block, press f7.  If the block is shorter than ten
seconds, the block will be played and you will be returned to the possition
in the file where you were when you pressed f7.
If the block is longer than ten seconds, you will not be returned to that
position directly.  You are then able to navigate inside the marked block
and can use the navigation keys, stop and start.  You cannot mark a block
within a block, so you can also not perform any task that require a marked
blok.  Press q to get back to the position in the main file where you were
when you pressed f7.

.SH Writing a block to an external audio file:
Press f8 to write the marked block to an external audio file.  For reliable
opperation, only write out .wav files.  After pressing f8, you will be
prompted for the name of the file in which you want the block of audio to be
written.  It is important that you give the .wav extention when you type the
file name.
Once the block has been written to the file, you will be returned to the
position in the main file where you were before you have pressed f8.

.SH Removing a block of audio:
Press f9 to remove the marked block.  The block will be removed and you will
be returned to the position where you were before pressing f9.  If you were
at a position which fell inside the block area when you pressed f9, you will
be returned to the position where the block started.  After the block has
been removed, a block is no longer marked.

Moving a block of audio
Currently, moving a block is done using three opperations.
1.  Mark the block.
2.  Write the block to an external audio file.
3.  Insert the external file you have just written out at the desired
position.

.SH MARKING AND SPLITTING TRACKS
Press p at the position where you want the track to end.
Move to the position where you want the second track to end and press p
again.
You can mark as many tracks as you like.
Once all tracks are marked, press S to split the tracks.
Note, it is a capital s.  Pressing capital j will move to the last track
marked.  The tracks will be written out to files with the format track_n.wav
where n is a number.  If you have less than ten tracks, it would simply be a
number from 1 through 9.  If there are more tracks, leading zeros will be
added to the number.  Only mark and split tracks when you have finished all
insertions and deletions in the file as the track markers are not updated if
you edit the file.  When tracks are marked, a file with the track positions
is stored in a text file with the name of the file you are editing with a .pos extention added.
E.G. test.wav is edited and tracks are marked.  The positions are stored in
test.wav.pos.

.SH FINISHING THE EDIT SESSION
Press q to exit the program.  The backup file and the edited file will be on
your disk, but all temporary files will be deleted.  The files you have
written out using f8 will also remain on the disk.
Dae does not have an undo feature, so if you accedentally mess up your file,
exit using q and copy the backup file over the one you have edited and start
over.
You can also copy the messed up file to another file if parts of the file
can be used.  You then can copy the backup to the origional file and edit it
again, inserting parts of the messed up file after you have written out
those parts.
This sounds complicated, but the main thing to remember is that you can
recover from an error if need be.
If you have messed up your file, just do not start editing it again before
you did something to the backup file because if you just start dae again in
this situation, the backup file will be replaced with your messed up file
and you can lose data.


.SH RESOURCES
Dae requires the scripts catchkey, pickafile  and
getterm to be in the execution path.
It also requires various standard UNIX utilities, normally available under
any UNIX installation.
It also requires ecasound and python to be installed.
An audio tutorial for dae should have been installed in
/usr/local/share/dae/tutorial.
It is called dae_tutorial.ogg and can be played with a program like mplayer.
.SH DIAGNOSTICS
.SH SEE ALSO
pickafile (1)
sox (1)
aoss (1)
.SH COPYRIGHT 
Dae is Copyright (C) copyright Meraka Institute, 
 Meraka, CSIR, P.O. Box 395, Pretoria, 0001, South Africa 
EMAIL: Willem van der Walt wvdwalt@csir.co.za
Web: http://www.meraka.org.za
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published
by the Free Software Foundation; either version 3 of the License,
or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston,
MA 02111-1307 USA.

.SH BUGS
For best opperation, make sure that you can play multiple audio streams at
once.  If you are using a screen reader like speakup, this is important as
you might otherwise have speakup and dae fighting over who gets the turn to
make sound.
If your machine cannot be set up in this way, you can still use dae by
turning speakup off using the insert-enter key combination on the numeric
keypad.

Dae can use up to three times the disk space that is taken up by the file
you are editing.  The working files are deleted when you exit the program,
but the backup file remains.  You must remove that by hand.
Any files that you have created using f8 write block, also remains.

Dae does not check for disk space before each operation, so make sure you
have enough space.  You can lose data if the disk becomes full during some
operations although your backup file should be safe.

By default, dae uses the device called "/dev/dsp" for both sound input and
sound output.  On some machines this might perform poorly and on some
machines it might not work at all.
You can try running dae using the aoss program to work around this problem. 
For example:
$aoss dae myfile.wav
Alternatively, you can edit the dae file.  At the top of the file, you will
find the lines:
soundin="/dev/dsp"
soundout="/dev/dsp"
You can change that to another device, for example:
soundin="alsa"
soundout="alsa"
Python version 2.6 and later complains about some of the code used in the
ecasound python module, but everything still works without problems.  With
python 2.6 you are likely to see these errors apon entering and exiting dae.
It is possible to get rid of these errors under python 2.6 and later by
compiling ecasound against your installed python.
To do that, do the following:
Steps are for Debean/ubuntu
1.  Install python-dev
    sudo apt-get install python-dev

2.  Get the source code for ecasound.
    apt-get source ecasound
3.  Change into the directory created by the previous step.
cd ecasound2.2-2.7.0
4.  Configure ecasound to be built for your python:
    ./configure --enable-pyecasound=c
5.  Build and install ecasound:
    make&&sudo make install


Although dae needs more work, I have used it for a number of years for all my
audio editing and had no major problems.

FUTURE PLANS FOR DAE
Add support for editing/writing compressed files like mp3 and ogg.
Add a save option where you can save to another file while still editing. 
This is currently possible by marking all of the file as a block and saving
it under another name using write block.
Add support for ladspa and ecasound plugins to be applied on parts of the
file and saved as part of the file afterwards.

.SH AUTHOR
Willem van der Walt
E-mail wvdwalt@csir.co.za


dae/dae/historic_perl_version/dae.pl
#!/usr/bin/perl -w
system "cp $ARGV[0] $ARGV[0]~";
$Main::bs = undef;
$Main::be = undef;
$Main::changed = 0;

$Main::soort = substr($ARGV[0],length($ARGV[0])-4,4);
$Main::infile = $ARGV[0];
$Main::encoded = 0;
if ($Main::soort eq ".mp3" or $Main::soort eq ".ogg")
{
  $Main::encoded = 1;
  }
  

use ExtUtils::testlib;

use Term::ReadKey;

use Audio::Ecasound qw(:simple :iam);
if ($Main::encoded eq 1)
{
  &convert_first;
  }
  

use strict;
# :iam is nicer without strict 'subs'
no strict 'subs';
on_error('');

# no strict 'subs' lets you do this:
eci("cs_add play_chainsetup
c_add chain1");
eci("-i:$Main::infile
        -o:/dev/dsp
cs_connect"); 

eci("ai-select $Main::infile");

$Main::masterformat = eci("ai-get-format");

 
    
eci("start");
use lib "./blib/lib";
use Term::Screen::ReadLine;

$Term::Screen::ReadLine::scr = new Term::Screen::ReadLine;

# test getch
# clear buffer out
$Term::Screen::ReadLine::scr->flush_input();
$Term::Screen::ReadLine::scr->at(25,0);
printf "master format is: %s ",$Main::masterformat;
$Term::Screen::ReadLine::scr->at(21,0)->puts("Digital audio, Enter Key (q to quit): ")->at(21,40);


$Term::Screen::ReadLine::ch = '';
while(($Term::Screen::ReadLine::ch = $Term::Screen::ReadLine::scr->getch()) ne 'q') 
{
       $Main::keepme = eci("getpos");

if ($Term::Screen::ReadLine::ch eq 'k1')
{
print "stop";
  eci("stop");
}
if ($Term::Screen::ReadLine::ch eq 'k2')
{
print "play";
eci("start");
}
if ($Term::Screen::ReadLine::ch eq 'kl')
{
&move_around;
}

if ($Term::Screen::ReadLine::ch eq 'kr')
{
  &move_around;
}
if ($Term::Screen::ReadLine::ch eq 'k3')
{
print "fast rewind";
eci("rewind 60");
}
if ($Term::Screen::ReadLine::ch eq 'k4')
{
print "fast forward";
eci("forward 60");
}
if ($Term::Screen::ReadLine::ch eq 'k5')
{

  $Main::bs = eci("getpos");
  
  printf "marked block startat : %d", $Main::bs;
}
if ($Term::Screen::ReadLine::ch eq 'k6')
{
$Main::be = eci("getpos");
printf "marked block end at : %d", $Main::be;
}

if ($Term::Screen::ReadLine::ch eq 'k7')
{
   $Main::vra = 0;
   $Main::blockout = "block$Main::soort";
  &writeblock;
printf "played %d second block", $Main::time;
}

if ($Term::Screen::ReadLine::ch eq 'k8')
{
    printf "\n file name to write block as? ";
  $Main::vra = 1;
  &writeblock;
printf "written %d second block to %s ", $Main::time, $Main::blockout;
}
if ($Term::Screen::ReadLine::ch eq 'k9')
{
  &delblock;

}
if ($Term::Screen::ReadLine::ch eq 'k10')
{
&record;
print "insert real-time";
}
if ($Term::Screen::ReadLine::ch eq 'k11')
{
  Term::ReadKey::ReadMode(1);
  system "pickafile";
    Term::ReadKey::ReadMode(0);
    $Main::fromfile =  `cat \$HOME/.kies/.\`getterm\`file`;
        chomp $Main::fromfile;
        &check_for_encoded;
        &record;
printf "inserted file : %s ",$Main::fromfile;
}
if ($Term::Screen::ReadLine::ch eq 'k12')
{
  $Main::einde = eci("cs-get-length");
    eci("setpos $Main::einde");
printf "go to end at : %d", $Main::einde;
}
if ($Term::Screen::ReadLine::ch eq 'kh')
{
    eci("setpos 0");
printf "go to start";
}
if ($Term::Screen::ReadLine::ch eq 'w')
{
printf "now at : %f ",$Main::keepme;
}
if ($Term::Screen::ReadLine::ch eq 'j')
{
printf "jump to? ";
    $Main::jump = $Term::Screen::ReadLine::scr->readline;
        eci("setpos $Main::jump");
}
if ($Term::Screen::ReadLine::ch eq 'h')
{
  $Term::Screen::ReadLine::scr->at(25,1)->puts("f1 stop f2 play f3 fr f4 ff f5 begin-block f6 end-block f7 play-block f8 write-block f9 del-block f10 record f11 ins-file f12 goto end");
}



      $Term::Screen::ReadLine::scr->at(21,40);
}

$Term::Screen::ReadLine::scr->at(22,0);


#stop;
eci("cs_disconnect");
eci("cs-remove default");


if ($Main::encoded eq 1 and $Main::changed eq 1)
{
   &convert_back;
   }
print "exiting.";
   
sub convert_first
{
printf "decoding %s first, please wait.",$Main::infile;
if ($Main::soort eq ".ogg")
{
  system "ogg123 $Main::infile -dwav -f editme.wav";
}
if ($Main::soort eq ".mp3")
{
system "mpg123 $Main::infile -w editme.wav";
}
$Main::orig_soort = $Main::soort;


  $Main::soort = ".wav";
$Main::infile = "editme.wav";
}

sub convert_back
{
printf "encoding %s first, please wait.",$ARGV[0];
if ($Main::orig_soort eq ".ogg")
{
  system "oggenc $Main::infile -o $ARGV[0]";
  }
  if ($Main::orig_soort eq ".mp3")
  {
    system "lame $Main::infile $ARGV[0]";
    }
unlink $Main::infile;


}


sub check_for_encoded

{
$Main::fromfiletype = substr($Main::fromfile,length($Main::fromfile)-4,4);

printf "decoding %s first, please wait.",$Main::fromfile;
if ($Main::fromfiletype eq ".ogg")
{
  system "ogg123 $Main::fromfile -dwav -f realtime.wav";
}
if ($Main::fromfiletype eq ".mp3")
{
system "mpg123 $Main::fromfile -w realtime.wav";
}
if ($Main::fromfiletype eq ".ogg" or $Main::fromfiletype eq ".mp3")
{
$Main::fromfile = "realtime.wav";
print "in if file type is : ",$Main::fromfiletype," from file is ",$Main::fromfile,"\n";

}
print "fromfile is ",$Main::fromfile;


}

sub delblock
{
if (!defined($Main::bs))
{
   printf "no block start defined\n";
       return;
       }
if (!defined($Main::be))
{
   printf "no block end defined\n";
       return;
       }
       
$Main::changed = 1;

unlink "l$Main::soort";
unlink "r$Main::soort";

$Main::time = $Main::be - $Main::bs;

eci("stop
 cs-disconnect default
cs-remove default");
#create chain to grab left side
eci("cs-add default
-i:$Main::infile
-o:l$Main::soort -t:$Main::bs
cs-connect
run");
eci("cs-disconnect default
cs-remove default");

#create chain to grab right side
eci("cs-add default
-i:$Main::infile -y:$Main::be
-o:r$Main::soort
cs-connect
run");
eci("cs-disconnect default
cs-remove default");


#restore modified file
unlink $Main::infile;
#create chain to concat left and right sides
eci("cs-add default
-i:r$Main::soort
-o:l$Main::soort -y:$Main::bs
cs-connect
run");
while (eci("engine-status") eq "running")
{
#wait
}

eci("cs-disconnect default
cs-remove default");
rename "l$Main::soort", $Main::infile;
$Main::bs = undef;
$Main::be = undef;
printf "deleted %d second block", $Main::time;
&playagain;
}

sub writeblock
{
if (!defined($Main::bs))
{
   printf "no block start defined\n";
       return;
       }
if (!defined($Main::be))
{
   printf "no block end defined\n";
       return;
       }

if ($Main::vra eq 1) {
  $Main::blockout = $Term::Screen::ReadLine::scr->readline;
  }


$Main::time = $Main::be - $Main::bs;

eci("stop
 cs-disconnect default
cs-remove default");
unlink "block$Main::soort";

#create chain to write block
eci("cs-add default
-i:$Main::infile -y:$Main::bs
-o:$Main::blockout -t:$Main::time
cs-connect");
eci("run");
if ($Main::vra ne 1)
  {
eci("cs-disconnect
  cs-remove default
  cs-add default
  -i:block$Main::soort
  -o:/dev/dsp
  cs-connect default
  start");
  
       
$Term::Screen::ReadLine::ch = '';
while(($Term::Screen::ReadLine::ch = $Term::Screen::ReadLine::scr->getch()) ne 'q') 
{
if ($Term::Screen::ReadLine::ch eq 'k1')
{
print "stop";
  eci("stop");
}
if ($Term::Screen::ReadLine::ch eq 'k2')
{
print "play";
eci("start");
}
if ($Term::Screen::ReadLine::ch eq 'kl')
{
&move_around;
}

if ($Term::Screen::ReadLine::ch eq 'kr')
{
  &move_around;
}
if ($Term::Screen::ReadLine::ch eq 'k3')
{
print "fast rewind";
eci("rewind 60");
}
if ($Term::Screen::ReadLine::ch eq 'k4')
{
print "fast forward";
eci("forward 60");
}
if ($Term::Screen::ReadLine::ch eq 'k12')
{
  $Main::einde = eci("cs-get-length");
    eci("setpos $Main::einde");
printf "go to end at : %d", $Main::einde;
}

      $Term::Screen::ReadLine::scr->at(21,40);
}

       }
       
eci("cs-disconnect default
cs-remove default");
&playagain;

}

sub record
{
$Main::changed = 1;

unlink "l$Main::soort";
unlink "r$Main::soort";

eci("stop");
$Main::plek = eci("getpos");

eci("cs-disconnect default
cs-remove default");
#create chain to grab left side
eci("cs-add default
-i:$Main::infile
-o:l$Main::soort -t:$Main::plek
cs-connect
run");
eci("cs-disconnect default
cs-remove default");

#create chain to grab right side
eci("cs-add default
-i:$Main::infile -y:$Main::plek
-o:r$Main::soort
cs-connect
run");
eci("cs-disconnect default
cs-remove default");
if ($Term::Screen::ReadLine::ch eq "k10")
  {
  
#record from soundcard
    unlink "realtime$Main::soort";
eci("cs-add default
-f:$Main::masterformat
-i:/dev/dsp
-o:realtime$Main::soort -ea:900
cs-connect");

$Term::Screen::ReadLine::ch = '';
while(($Term::Screen::ReadLine::ch = $Term::Screen::ReadLine::scr->getch()) ne 'q') 
{
if ($Term::Screen::ReadLine::ch eq 'k1')
{
print "stop";
  eci("stop");
}
if ($Term::Screen::ReadLine::ch eq 'k2')
{
print "start";
eci("start");
}
}
$Term::Screen::ReadLine::ch = "";
eci("cs-disconnect
cs-remove default");
    $Main::fromfile = "realtime$Main::soort";
    
}

#restore modified file
unlink $Main::infile;
#create chain to concat left, recorded and right files
 
eci("cs-add default
-f:$Main::masterformat
-i:resample-hq,auto,$Main::fromfile
-o:l$Main::soort -y:$Main::plek
cs-connect
run");
$Main::plek = $Main::plek + eci("getpos");

eci("cs-disconnect default
cs-remove default");
 
eci("cs-add default
-i:r$Main::soort
-o:l$Main::soort -y:$Main::plek
cs-connect
run
cs-disconnect
cs-remove default");

rename "l$Main::soort", $Main::infile;
&playagain;
}

sub playagain 
{

eci("cs-add default
-i:$Main::infile
-o:/dev/dsp
cs-connect default
setpos $Main::keepme
start");

}

sub move_around
{
$Main::move_count = 1;
$Main::engine = eci("engine-status");
if ($Main::engine eq 'running')
{
  eci("stop");
  }
if ($Term::Screen::ReadLine::ch eq 'kl')
{
  $Main::direction = 'rewinding';
  }
  else
  {
    $Main::direction = 'forwarding';
    }
    
  
while($Term::Screen::ReadLine::scr->key_pressed(1)) 
{
$Main::move_count = $Main::move_count + 1;
$Term::Screen::ReadLine::ch = $Term::Screen::ReadLine::scr->getch();
}
if ($Main::direction eq 'rewinding')
{
  eci("rewind $Main::move_count");
  }
  else 
  {
    eci("forward $Main::move_count");
    }
 printf "%s %d seconds",$Main::direction,$Main::move_count;    
    
if ($Main::engine eq 'running')
{
eci("start");
}

}
